src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListenerBase.cs

```diff
@@ -272,22 +286,123 @@ private static async Task ProcessWorkCoreAsync(ImmutableArray<Work> work, Stream
+        var projectPath = Path.GetDirectoryName(project.FilePath);
+        if (projectPath is null)
         {
-            logger.LogTrace("Skipped writing data for {projectId}", project.Id);
+            _logger.LogInformation("projectPath is null, skip update for {projectId}", project.Id);
             return;
         }
 
-        stream.WriteProjectInfoAction(ProjectInfoAction.Update);
-        await stream.WriteProjectInfoAsync(projectInfo, cancellationToken).ConfigureAwait(false);
+        var intermediateOutputPath = Path.GetDirectoryName(project.CompilationOutputInfo.AssemblyPath);
+        if (intermediateOutputPath is null)
+        {
+            _logger.LogInformation("intermediateOutputPath is null, skip update for {projectId}", project.Id);
+            return;
+        }
+
+        if (!_projectEntryMap.TryGetValue(project.Id, out var entry))
+        {
+            _projectEntryMap[project.Id] = entry = new();
+        }
+
+        var (didChange, projectInfo) = await DidProjectInfoChangeAsync(project, entry, projectPath, intermediateOutputPath, cancellationToken).ConfigureAwait(false);
+        if (didChange)
+        {
+            projectInfo.AssumeNotNull();
+            UpdateEntry(project, projectInfo, entry);
+
+            stream.WriteProjectInfoAction(ProjectInfoAction.Update);
+            await stream.WriteProjectInfoAsync(projectInfo, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    private void UpdateEntry(Project project, RazorProjectInfo projectInfo, ProjectEntry entry)
+    {
+        entry.RootNamespace = projectInfo.RootNamespace;
+        entry.ConfigurationHash = projectInfo.Configuration.GetHashCode();
+        entry.DocumentsHash = projectInfo.Documents.GetHashCode();
+
+        _cachedTagHelperResolver.TryGetId(project.Id, out var resultId);
+        entry.TagHelpersResultId = resultId;
+    }
+
+    private async Task<(bool, RazorProjectInfo? projectInfo)> DidProjectInfoChangeAsync(Project project, ProjectEntry entry, string projectPath, string intermediateOutputPath, CancellationToken cancellationToken)
+    {
+        ImmutableArray<DocumentSnapshotHandle> documents = default;
+        var csharpLanguageVersion = (project.ParseOptions as CSharpParseOptions)?.LanguageVersion ?? LanguageVersion.Default;
+        CachedTagHelperResolver.DeltaResult? delta = null;
+
+        var configuration = RazorProjectInfoHelpers.ComputeRazorConfigurationOptions(project.AnalyzerOptions.AnalyzerConfigOptionsProvider, out var rootNamespace);
+        if (configuration is null)
+        {
+            return (false, null);
+        }
+
+        if (entry.RootNamespace is null)
+        {
+            return await GetChangedValueAsync().ConfigureAwait(false);
+        }
+
+        if (!entry.RootNamespace.Equals(rootNamespace))
+        {
+            return await GetChangedValueAsync().ConfigureAwait(false);
+        }
+
+        if (entry.ConfigurationHash != configuration.GetHashCode())
```

Comment:
This worries me a bit. Hash codes are not good enough to guarantee equality or inequality. Please use `Microsoft.AspNetCore.Razor.Utilities.Checksum` to produce a SHA256 that provides much stronger identity.